extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -59,34 +70,73 @@ private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity security
                     public CompletionStage<Boolean> apply(Permission permission) {
                         return securityIdentity.checkPermission(permission);
                     }
-                })
-                .build();
+                }).build();
     }
 
     public CompletionStage<SecurityIdentity> authenticate(RoutingContext context,
             IdentityProviderManager identityProviderManager,
             DefaultTenantConfigResolver resolver) {
         Cookie sessionCookie = context.request().getCookie(SESSION_COOKIE_NAME);
+        TenantConfigContext configContext = resolver.resolve(context, true);
 
         // if session already established, try to re-authenticate
         if (sessionCookie != null) {
             String[] tokens = sessionCookie.getValue().split(COOKIE_DELIM);
-            return authenticate(identityProviderManager, new IdTokenCredential(tokens[0], context))
-                    .thenCompose(new Function<SecurityIdentity, CompletionStage<SecurityIdentity>>() {
+            String idToken = tokens[0];
+            String accessToken = tokens[1];
+            String refreshToken = tokens[2];
+
+            return authenticate(identityProviderManager, new IdTokenCredential(idToken, context))
+                    .handle(new BiFunction<SecurityIdentity, Throwable, SecurityIdentity>() {
                         @Override
-                        public CompletionStage<SecurityIdentity> apply(SecurityIdentity securityIdentity) {
-                            return CompletableFuture
-                                    .completedFuture(augmentIdentity(securityIdentity, tokens[1], tokens[2], context));
+                        public SecurityIdentity apply(SecurityIdentity identity, Throwable throwable) {
+                            if (isLogout(context, configContext)) {
```

Comment:
It looks like this check should after a throwable check, otherwise we will be initiating a logout flow even if the token signature has failed for example, but the user session must have a valid token (except for the possible expiry)

Reply:
Yeah, will change to only do logout if identity is available.